File: app/src/main/cpp/src/Player.cpp
//
// Created by Pavel Konev on 11/15/20.
//

#include "Player.h"

namespace avtools
{
    /*static*/ std::unique_ptr<Player> Player::create(std::string file_path, ErrorCode &error_code) {
        return std::unique_ptr<Player>();
    }

    Player::Player() {
    }

    Player::~Player() {

    }

    Player::StateId Player::get_state_id() const {
        return Player::StateId::Error;
    }

    FileDuration Player::get_duration() const {
        return 0;
    }

    PlaybackPosition Player::get_position() const {
        return 0;
    }

    PlaybackSpeed Player::get_speed() const {
        return 0;
    }

    VolumeLevel Player::get_volume_level() const {
        return 0;
    }

    ErrorCode Player::set_position(PlaybackPosition position) {
        return ErrorCode::NotSupported;
    }

    ErrorCode Player::set_speed(PlaybackSpeed speed) {
        return ErrorCode::NotSupported;
    }

    ErrorCode Player::set_volume_level(VolumeLevel volume_level) {
        return ErrorCode::NotSupported;
    }

    ErrorCode Player::start() {
        return ErrorCode::NotSupported;
    }

    void Player::stop() {

    }
}

File: app/src/main/java/com/avtools/player/Player.java
//
// Created by Pavel Konev on 11/6/20.
//

package com.avtools.player;

public class Player
{
    public enum StateId
    {
        NotInitialized,
        Stopped,
        Started,
        Paused,
        Error
    };

    public enum ErrorCode
    {
        NoError,
        NotInitialized,
        UnknownFormat,
        InvalidArgs,
        OutOfBounds,
        ResourceBusy,
        NotSupported,
        UnknownError
    };

    public class State
    {
        private State(StateId idArg, float durationArg, float positionArg, float speedArg, float volumeArg)
        {
            id       = idArg;
            duration = durationArg;
            position = positionArg;
            speed    = speedArg;
            volume   = volumeArg;
        }

        public StateId id;
        public float   duration;
        public float   position;
        public float   speed;
        public float   volume;
    };

    /*
     * Initialize AV player.
     * @param filePath file path
     */
    public native ErrorCode init(String filePath);

    /*
     * Destroy AV player.
     */
    public native void uninit();

    /**
     * Start AV player.
     * @return error code
     */
    public native ErrorCode start();

    /**
     * Stop AV player.
     */
    public native void stop();

    /**
     * Set playback speed (1x supported only).
     * @return error code
     */
    public native ErrorCode setSpeed(float speed);

    /**
     * Set playback position (0-duration).
     * @return error code
     */
    public native ErrorCode setPosition(float position);

    /**
     * Set volume (0-1).
     * @return error code
     */
    public native ErrorCode setVolume(float volume);

    /**
     * Get player state.
     * @return state
     */
    public native State getState();
}
